crates/bevy_ecs/src/storage/table/column.rs

```diff
@@ -0,0 +1,705 @@
+    /// # Safety
+    /// - `T` must match the type of data that's stored in this [`ThinColumn`]
+    /// - `len` must match the actual length of this column (number of elements stored)
+    pub unsafe fn get_data_slice_for<T>(&self, len: usize) -> &[UnsafeCell<T>] {
+        self.data.get_sub_slice(len)
+    }
+
+    /// Get a slice to the added [`ticks`](Tick) in this [`ThinColumn`].
+    ///
+    /// # Safety
+    /// - `len` must match the actual length of this column (number of elements stored)
+    pub unsafe fn get_added_ticks_slice(&self, len: usize) -> &[UnsafeCell<Tick>] {
+        self.added_ticks.as_slice(len)
+    }
+
+    /// Get a slice to the changed [`ticks`](Tick) in this [`ThinColumn`].
+    ///
+    /// # Safety
+    /// - `len` must match the actual length of this column (number of elements stored)
+    pub unsafe fn get_changed_ticks_slice(&self, len: usize) -> &[UnsafeCell<Tick>] {
+        self.changed_ticks.as_slice(len)
+    }
+
+    /// Get a slice to the calling locations that last changed each value in this [`ThinColumn`]
+    ///
+    /// # Safety
+    /// - `len` must match the actual length of this column (number of elements stored)
+    #[cfg(feature = "track_change_detection")]
+    pub unsafe fn get_changed_by_slice(
+        &self,
+        len: usize,
+    ) -> &[UnsafeCell<&'static Location<'static>>] {
+        self.changed_by.as_slice(len)
+    }
+
+    /// Fetches the calling location that last changed the value at `row`.
+    ///
+    /// # Safety
+    /// - `row` must be in bounds
+    ///
+    /// Note: The values stored within are [`UnsafeCell`].
+    /// Users of this API must ensure that accesses to each individual element
+    /// adhere to the safety invariants of [`UnsafeCell`].
+    #[inline]
+    #[cfg(feature = "track_change_detection")]
+    pub unsafe fn get_changed_by_unchecked(
+        &self,
+        row: TableRow,
+    ) -> &UnsafeCell<&'static Location<'static>> {
+        self.changed_by.get_unchecked(row.as_usize())
+    }
+}
+
+/// A type-erased contiguous container for data of a homogeneous type.
+///
+/// Conceptually, a [`Column`] is very similar to a type-erased `Vec<T>`.
+/// It also stores the change detection ticks for its components, kept in two separate
+/// contiguous buffers internally. An element shares its data across these buffers by using the
+/// same index (i.e. the entity at row 3 has it's data at index 3 and its change detection ticks at index 3).
+///
+/// Like many other low-level storage types, [`Column`] has a limited and highly unsafe
+/// interface. It's highly advised to use higher level types and their safe abstractions
+/// instead of working directly with [`Column`].
+#[derive(Debug)]
+pub struct Column {
+    pub(super) data: BlobVec,
+    pub(super) added_ticks: Vec<UnsafeCell<Tick>>,
+    pub(super) changed_ticks: Vec<UnsafeCell<Tick>>,
+    #[cfg(feature = "track_change_detection")]
+    changed_by: Vec<UnsafeCell<&'static Location<'static>>>,
+}
```

Comment:
Is there a reason this isn't implemented as a `ThinColumn` along with a length and capacity `usize`s?

Reply:
`Column` will be completely replaced by `ThinColumn` in a follow-up PR (see #14824)

Reply:
Seems reasonable then